coalib/core/Bear.py

```diff
@@ -0,0 +1,469 @@
+from coalib.settings.Section import Section
+
+
+class Bear:
+    """
+    A bear contains the actual subroutine that is responsible for checking
+    source code for certain specifications. However, it can actually do
+    whatever it wants with the files it gets.
+
+    This is the base class for every bear. If you want to write a bear, you
+    will probably want to look at the ``ProjectBear`` and ``FileBear`` classes
+    that inherit from this class.
+
+    To indicate which languages your bear supports, just give it the
+    ``LANGUAGES`` value which should be a set of string(s):
+
+    >>> class SomeBear(Bear):
+    ...     LANGUAGES = {'C', 'CPP', 'C#', 'D'}
+
+    To indicate the requirements of the bear, assign ``REQUIREMENTS`` a set
+    with instances of ``PackageRequirements``.
+
+    >>> class SomeBear(Bear):
+    ...     REQUIREMENTS = {
+    ...         PackageRequirement('pip', 'coala_decorators', '0.2.1')}
+
+    If your bear uses requirements from a manager we have a subclass from,
+    you can use the subclass, such as ``PipRequirement``, without specifying
+    manager:
+
+    >>> class SomeBear(Bear):
+    ...     REQUIREMENTS = {PipRequirement('coala_decorators', '0.2.1')}
+
+    To specify additional attributes to your bear, use the following:
+
+    >>> class SomeBear(Bear):
+    ...     AUTHORS = {'Jon Snow'}
+    ...     AUTHORS_EMAILS = {'jon_snow@gmail.com'}
+    ...     MAINTAINERS = {'Catelyn Stark'}
+    ...     MAINTAINERS_EMAILS = {'catelyn_stark@gmail.com'}
+    ...     LICENSE = 'AGPL-3.0'
+    ...     ASCIINEMA_URL = 'https://asciinema.org/a/80761'
+
+    If the maintainers are the same as the authors, they can be omitted:
+
+    >>> class SomeBear(Bear):
+    ...     AUTHORS = {'Jon Snow'}
+    ...     AUTHORS_EMAILS = {'jon_snow@gmail.com'}
+    >>> SomeBear.maintainers
+    {'Jon Snow'}
+    >>> SomeBear.maintainers_emails
+    {'jon_snow@gmail.com'}
+
+    If your bear needs to include local files, then specify it giving strings
+    containing relative file paths to the INCLUDE_LOCAL_FILES set:
+
+    >>> class SomeBear(Bear):
+    ...     INCLUDE_LOCAL_FILES = {'checkstyle.jar', 'google_checks.xml'}
+
+    To keep track easier of what a bear can do, simply tell it to the
+    ``CAN_FIX`` and the ``CAN_DETECT`` sets. Possible values are:
+
+    >>> CAN_DETECT = {'Syntax', 'Formatting', 'Security', 'Complexity',
+    ... 'Smell', 'Unused Code', 'Redundancy', 'Variable Misuse', 'Spelling',
+    ... 'Memory Leak', 'Documentation', 'Duplication', 'Commented Code',
+    ... 'Grammar', 'Missing Import', 'Unreachable Code', 'Undefined Element',
+    ... 'Code Simplification'}
+    >>> CAN_FIX = {'Syntax', ...}
+
+    Specifying something to ``CAN_FIX`` makes it obvious that it can be
+    detected too, so it may be omitted:
+
+    >>> class SomeBear(Bear):
+    ...     CAN_DETECT = {'Syntax', 'Security'}
+    ...     CAN_FIX = {'Redundancy'}
+    >>> list(sorted(SomeBear.can_detect))
```

Comment:
since sorted is O(nlogn) or so anyway there's no benefit in making it an iterator